kong/runloop/handler.lua

```diff
@@ -417,7 +439,31 @@ local function register_events()
 
   if db.strategy == "off" then
     worker_events.register(function()
+      local ok = kong.worker_events.post("crud", "upstreams", {
+        operation = "delete",
+        entity = "all",
+      })
+      if not ok then
+        kong.log.err("failed posting invalidation event for all upstreams")
+      end
```

Comment:
I may be wrong, but I've been thinking this: if `n` workers get a `flip_config` event, and then each of them posts a "delete all" event, won't this cause `n * n` "delete all" events in total?

(Maybe that's what's happening with `balancer` events nested inside `crud` events, which caused the burst to be extra problematic? I have vague memories about this.)

I'm now thinking that it might be more efficient to post these events from `declarative.load_into_cache_with_events` instead to avoid the O(n^2), and post them directly to `worker_events.post("balancer", "upstreams", ...)` bypassing `"crud"` altogether (the caches are clear on a DB config flip anyway).

I was even going to suggest calling `balancer.on_upstream_event` directly here, but at least for the creation of balancers is good to do them async (as some incur in DNS delays; in `kong.init`, `balancer.init` happens inside a timer for this reason), so posting `"balancer"` worker events from `declarative.load_into_cache_with_events` is probably the better option.

(Also, non-blocker side comment: I know it's a matter of taste and my love of types :sweat_smile:, but I'd probably make `delete_all` a new operation for `balancer.on_upstream_event` instead of making `"all"` a magic `upstream_data` value)

Reply:
It makes sense, I made the changes to fix that. Also, I think you're right about the magic `upstream_data` 😳. Also fixed that